### DISCUSSION TITLE
experimental.esmExternals feedback

### DISCUSSION THREAD

**Discussion by timneutkens** [vercel/next.js#27876] [Ideas]
**Update: This has shipped! https://nextjs.org/blog/next-12#es-modules-support-and-url-imports**

---

Thread to collect feedback about `experimental.esmExternals: true`.

**Comment by GrossDesignCo** [comment]
So call me dumb, I: 

1. Updated to 11.1.0
2. changed my next.config.js from:

```
const withTM = require('next-transpile-modules')([
 // list of modules
]);

const withPlugins = require('next-compose-plugins');

const nextConfig = {
  modern: true,
};

module.exports = withPlugins([withTM], nextConfig);
```

to:
```
const nextConfig = {
  modern: true,
  experimental: { esmExternals: true },
};

module.exports = nextConfig;
```

And it throws the classic:
```
SyntaxError: Unexpected token 'export'
    at wrapSafe (internal/modules/cjs/loader.js:979:16)
    at Module._compile (internal/modules/cjs/loader.js:1027:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
```

Am I missing a piece of the workflow?

**Reply by deadcoder0904** [reply]
Same error as @GrossDesignCo & I followed the same thing.

```bash
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: my-app\node_modules\unist-util-visit\index.js
require() of ES modules is not supported.
require() of my-app\node_modules\unist-util-visit\index.js from my-app\rehype\highlight-code.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from my-app\node_modules\unist-util-visit\package.json.

    at new NodeError (node:internal/errors:363:5)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1126:13)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at Module.require (node:internal/modules/cjs/loader:1013:19)
    at require (node:internal/modules/cjs/helpers:93:18)
    at Object.<anonymous> (my-app\rehype\highlight-code.js:4:15)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32) {
  code: 'ERR_REQUIRE_ESM'
```

**Reply by maxquinn** [reply]
Getting the same issue on vercel deployment. Next build works fine locally 
```
22:48:18.877 | > Build error occurred
-- | --
22:48:18.880 | Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /vercel/path0/node_modules/remark/index.js
22:48:18.880 | require() of ES modules is not supported.
22:48:18.880 | require() of /vercel/path0/node_modules/remark/index.js from /vercel/path0/.next/server/pages/blog/posts/[slug].js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
22:48:18.880 | Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /vercel/path0/node_modules/remark/package.json.
22:48:18.883 | at Object.Module._extensions..js (internal/modules/cjs/loader.js:1102:13)
22:48:18.883 | at Module.load (internal/modules/cjs/loader.js:950:32)
22:48:18.883 | at Function.Module._load (internal/modules/cjs/loader.js:790:14)
22:48:18.883 | at Module.require (internal/modules/cjs/loader.js:974:19)
22:48:18.883 | at require (internal/modules/cjs/helpers.js:92:18)
22:48:18.883 | at Object.338 (/vercel/path0/.next/server/pages/blog/posts/[slug].js:37:41)
22:48:18.883 | at __webpack_require__ (/vercel/path0/.next/server/webpack-runtime.js:25:42)
22:48:18.883 | at __webpack_exec__ (/vercel/path0/.next/server/pages/blog/posts/[slug].js:301:39)
22:48:18.883 | at /vercel/path0/.next/server/pages/blog/posts/[slug].js:302:78
22:48:18.883 | at Function.__webpack_require__.X (/vercel/path0/.next/server/webpack-runtime.js:108:21) {
22:48:18.884 | type: 'NodeError',
22:48:18.884 | code: 'ERR_REQUIRE_ESM'
22:48:18.884 | }
```

**Reply by timneutkens** [reply]
Please provide a full reproduction in a repository.

**Reply by deadcoder0904** [reply]
https://github.com/deadcoder0904/next-esm-error

**Reply by GrossDesignCo** [reply]
A little self-learning here, I updated our dependencies to use `"type": "module"` and they largely started to work with the above config change!

The only issue I'm having now is that directory imports don't work (which are only supported in Node via the `--experimental-specifier-resolution=node` flag.

Is there any way to pass that flag into Next.js's build process?

Edit, I'm assuming that's the issue based on this error message:

```
info  - Checking validity of types  
info  - Creating an optimized production build  
info  - Compiled successfully

> Build error occurred
Error [ERR_UNSUPPORTED_DIR_IMPORT]: Directory import '[my-url]' is not supported resolving ES modules imported from [my-url]
    at finalizeResolution (internal/modules/esm/resolve.js:281:17)
    at moduleResolve (internal/modules/esm/resolve.js:708:10)
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:819:11)
    at Loader.resolve (internal/modules/esm/loader.js:89:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:242:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:73:40)
    at link (internal/modules/esm/module_job.js:72:36) {
  type: 'NodeError',
  code: 'ERR_UNSUPPORTED_DIR_IMPORT',
  url: [my-url]
}
```

Edit, maybe I'm crazy, it did start working, but next I'm getting a hooks error. There's got to be something funky with our internal react component lib.

**Reply by kimbaudi** [reply]
@GrossDesignCo - isn't `modern: true` deprecated? https://github.com/vercel/next.js/pull/19275

**Reply by kimbaudi** [reply]
@GrossDesignCo - I don't think enabling `experimental.esmExternals: true` flag means you can just stop using `next-transpile-modules`.

**Reply by eric-burel** [reply]
Hi, do I need Webpack 5 though for this to work? In addition to Next 12. Because I still hit  "ERR_REQUIRE_ESM" in a package I try to build

**Reply by The-Code-Monkey** [reply]
@GrossDesignCo Did you manage to fix this issue?

**Comment by ppbl** [comment]
I use custom server

```js
// next.config.js
module.exports = {
  experimental: { esmExternals: true }
}
```

```js
// package.json
{
  "type": "module"
}
```

```js
// server.js
import next from 'next'

const dev = process.env.NODE_ENV === 'development'
const app = next({ dev })
const handle = app.getRequestHandler()

app.prepare().then(() => {
  // custom server
})
```

Run `yarn dev`, It is equivalent to running `node server.js`

```
/Users/ppbl/xxx/xxx/node_modules/next/dist/server/config.js:345
        const userConfigModule = require(path);
                                 ^

Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/ppbl/xxx/xxx/next.config.js from /Users/ppbl/xxx/xxx/node_modules/next/dist/server/config.js not supported.
next.config.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
Instead rename next.config.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /Users/ppbl/xxx/xxx/package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).

    at Object.loadConfig [as default] (/Users/ppbl/xxx/xxx/node_modules/next/dist/server/config.js:345:34)
    at async NextServer.loadConfig (/Users/ppbl/xxx/xxx/node_modules/next/dist/server/next.js:112:22)
    at async NextServer.prepare (/Users/ppbl/xxx/xxx/node_modules/next/dist/server/next.js:94:24) {
  code: 'ERR_REQUIRE_ESM'
}
```

**Reply by timneutkens** [reply]
This has nothing to do with `esmExternals`. You're importing Next.js itself as an ES Module which is not supported yet.

**Comment by mikestopcontinues** [comment]
Importing from third-party esm seems to be working, but all of my `index.ts` files are now broke. E.G. `export * from './SubComponent';`. I have to import named exports for it to work. What am I missing?

**Reply by nghiepdev** [reply]
I have the same problem as you.

**Reply by timneutkens** [reply]
Please provide a full reproduction in a repository.

**Reply by nghiepdev** [reply]
@timneutkens 
https://github.com/nghiepit/next11.1-esm-bug/blob/main/lib/my-lib/utils.ts#L4

**Reply by bespoyasov** [reply]
Have the same problem. 

Seems to be related to ESM import in the submodules. Simple re-export `*` from `index.ts` works but if at least one submodule imports an ESM re-export breaks.

**Reply by bespoyasov** [reply]
Named re-export like `export {something} from './submodule';` works fine though.

**Comment by Henzelix** [comment]
I upgraded to Next.js 11.1.0 for support of ES modules, but I still get 'export' error:
I'm using `PrismAsyncLight` from`react-syntax-highlighter` like this:
```
import { PrismAsyncLight as SyntaxHighlighter } from 'react-syntax-highlighter';
import coldarkDark from 'react-syntax-highlighter/dist/esm/styles/prism/coldark-dark';
SyntaxHighlighter.supportedLanguages;
```
And I get error:
```
/home/mikolaj/Pulpit/classy-next/classy/node_modules/react-syntax-highlighter/dist/esm/styles/prism/coldark-dark.js:1
export default {
^^^^^^

SyntaxError: Unexpected token 'export'
```

**Reply by timneutkens** [reply]
Please provide a full reproduction in a repository.

**Reply by Henzelix** [reply]
@timneutkens unfortunatelly, I can't

**Reply by timneutkens** [reply]
Why not? You don't have to share application code.

**Reply by Henzelix** [reply]
@timneutkens So, can you explain what shoud I reproduce? Importing `import coldarkDark from 'react-syntax-highlighter/dist/esm/styles/prism/coldark-dark';` from package [`react-syntax-higlighter`](https://github.com/react-syntax-highlighter/react-syntax-highlighter) is causing the issue, because it's using ES modules. What code you would like to see then?

**Reply by Henzelix** [reply]
@timneutkens this is my `next.config.js`:
```
module.exports = {
  reactStrictMode: true,
  images: {
    domains: ['source.boringavatars.com', 'firebasestorage.googleapis.com'],
  },
  experimental: { esmExternals: true }
}
```

**Reply by sokra** [reply]
react-syntax-highlighter is missing `"type": "module"`, so node.js doesn't treat the files as ESM.

Best would be a `package.json` with `{ "type": "module": }` in `react-syntax-highlighter/dist/esm/package.json`.

**Reply by kimbaudi** [reply]
@Henzelix - your issue can be resolved by changing your import statement from `react-syntax-highlighter/dist/esm/styles/prism` to `react-syntax-highlighter/dist/cjs/styles/prism`. please take a look at https://github.com/react-syntax-highlighter/react-syntax-highlighter/issues/230 and possibly https://github.com/remarkjs/react-markdown/pull/645.

**Comment by sandcastle** [comment]
When enabling, we get the following error:

```
> Build error occurred
Error: Collecting page data for /about is still timing out after 2 attempts. See more info here https://nextjs.org/docs/messages/page-data-collection-timeout
    at onRestart (/Users/glenn/dev/x-platform/entertainment/node_modules/next/dist/build/index.js:429:31)
    at Worker1.isPageStatic (/Users/glenn/dev/x-platform/entertainment/node_modules/next/dist/lib/worker.js:49:40)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Span.traceAsyncFn (/Users/glenn/dev/x-platform/entertainment/node_modules/next/dist/telemetry/trace/trace.js:60:20)
    at async /Users/glenn/dev/x-platform/entertainment/node_modules/next/dist/build/index.js:481:48
    at async Span.traceAsyncFn (/Users/glenn/dev/x-platform/entertainment/node_modules/next/dist/telemetry/trace/trace.js:60:20)
    at async Promise.all (index 2)
    at async /Users/glenn/dev/x-platform/entertainment/node_modules/next/dist/build/index.js:466:13
    at async Span.traceAsyncFn (/Users/glenn/dev/x-platform/entertainment/node_modules/next/dist/telemetry/trace/trace.js:60:20)
    at async /Users/glenn/dev/x-platform/entertainment/node_modules/next/dist/build/index.js:421:127
    at async Span.traceAsyncFn (/Users/glenn/dev/x-platform/entertainment/node_modules/next/dist/telemetry/trace/trace.js:60:20)
```

Build log:

```
info  - Loaded env from /Users/glenn/dev/x-platform/entertainment/.env.production
info  - Loaded env from /Users/glenn/dev/x-platform/entertainment/.env
info  - Using webpack 5. Reason: Enabled by default https://nextjs.org/docs/messages/webpack5
warn  - You have enabled experimental feature(s).
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

info  - Checking validity of types  
info  - Using external babel configuration from /Users/glenn/dev/x-platform/entertainment/babel.config.js
info  - Using external babel configuration from /Users/glenn/dev/x-platform/entertainment/babel.config.js
info  - Using external babel configuration from /Users/glenn/dev/x-platform/entertainment/babel.config.js
info  - Using external babel configuration from /Users/glenn/dev/x-platform/entertainment/babel.config.js
info  - Using external babel configuration from /Users/glenn/dev/x-platform/entertainment/babel.config.js
info  - Using external babel configuration from /Users/glenn/dev/x-platform/entertainment/babel.config.js
info  - Using external babel configuration from /Users/glenn/dev/x-platform/entertainment/babel.config.js
info  - Using external babel configuration from /Users/glenn/dev/x-platform/entertainment/babel.config.js
info  - Creating an optimized production build  
info  - Compiled successfully
warn  - Restarted collecting page data for /_app because it took more than 60 seconds
warn  - See more info here https://nextjs.org/docs/messages/page-data-collection-timeout
warn  - Restarted collecting page data for /_app because it took more than 60 seconds
```

All I did was remove the transpile step for `@popperjs/core`, if I add this back - it all works

```js
// const withTM = require('next-transpile-modules')(['@popperjs/core']);

module.exports = withBundleAnalyzer(
  // withTM(
  { /*config*/ }
  // )
);
```

**Reply by kimbaudi** [reply]
@sandcastle - I don't think enabling `experimental.esmExternals: true` flag means you no longer need `next-transpile-modules`.

**Comment by wereHamster** [comment]
Enabling esmExternals should also transpile the dependencies via babel.

Chances are really good that modules which use ESM are also written in modern JS. While Next.js transpiles first-party code (btw, even with modern:true arrow functions and class syntax are downcompiled to ES5?), it won't touch third-party code and so any modern syntax is retained and will cause issues in older browsers.

Without this there will always be the need to carry next-transpile-modules along in our configurations.

Somewhat related issue is that it's still impossible to easily specify what browsers the project supports, as browserslist configuration is not respected (https://github.com/vercel/next.js/discussions/12826), and requires custom babelrc.

**Reply by kimbaudi** [reply]
@wereHamster - isn't `modern: true` deprecated? https://github.com/vercel/next.js/pull/19275

**Comment by Manubi** [comment]
Hi,
Faced a couple of errors today. One of them I posted here: https://github.com/maticzav/graphql-middleware/issues/433
I couldn't build the project anymore and nothing seemed to work.

**Comment by BlackGlory** [comment]
I get this error:

```
Server Error

Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.
This error happened while generating the page. Any console logs will be displayed in the terminal window.
Source

src/pages/_document.tsx (5:31) @ Function.getInitialProps

  3 | export default class MyDocument extends Document {
  4 |   static async getInitialProps(ctx: DocumentContext) {
> 5 |     const initialProps = await Document.getInitialProps(ctx)
    |                               ^
  6 |     return { ...initialProps }
  7 |   }
  8 | 
```

## Solution
As mentioned in the comments, please import the files of the ESM module separately, instead of importing the index file that contains them.

**Reply by sandcastle** [reply]
We had something similar - it was caused by exporting multiple files through an index file, we broke out each import into its own line with a full path and it resolved the issue

**Reply by GrossDesignCo** [reply]
I’m pretty sure this is due to node.js not supporting directory imports (yet). There’s only an experimental flag for it at the moment

**Reply by ivan-kleshnin** [reply]
> There’s only an experimental flag for it at the moment

Yep. Unfortunately the flag alone doesn't work `$ NODE_OPTIONS="--experimental-specifier-resolution=node" yarn dev` 

```
TypeError [ERR_INVALID_MODULE_SPECIFIER]: Invalid module "file:///Users/.../node_modules/next/dist/bin/next" 
```

**Comment by leerob** [comment]
If anyone wants to check out a working example, I've updated my site to use the latest remark/rehype packages (which are ESM only) using this new flag.

https://github.com/leerob/leerob.io

**Reply by deadcoder0904** [reply]
Hey, @leerob thanks for the example. I want to do something similar but I'd like to access file outside of `scripts/` but inside `src/utils` like https://github.com/deadcoder0904/better-code-blocks/blob/main/scripts/build-rss.js#L8-L9 but unable to do it with Next v11.1.0 as I get this error -> https://github.com/vercel/next.js/issues/28111

I tried your `.mjs` approach but I get:

> Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'my-app\src\utils\mdx' imported from my-app\scripts\build-sitemap.mjs

I tried using relative path in `tsconfig.json` like `@/utils/mdx` but it gives another error & even absolute path doesn't work.

Any ideas? Or do I have to duplicate the same code of looping through blogposts just to generate sitemap & rss? I'm out of ideas.

**Comment by Cyral** [comment]
With esmExternals enabled, d3 (which recently switched to `type: module`) in version 7 still cannot be loaded by Next.

```
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /Users/***/frontend/web/node_modules/d3-scale/src/index.js
require() of ES modules is not supported.
require() of /Users/***/frontend/web/node_modules/d3-scale/src/index.js from /Users/***/frontend/web/.next/server/pages/_app.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/***/frontend/web/node_modules/d3-scale/package.json.

    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1102:13)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:14)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at Object.d3-scale (/Users/***/frontend/web/.next/server/pages/_app.js:1352:18)
    at __webpack_require__ (/Users/***/frontend/web/.next/server/webpack-runtime.js:33:42)
    at eval (webpack-internal:///./components/***.tsx:9:66)
    at Object../components/strategy/StrategyGraph.tsx (/Users/***/frontend/web/.next/server/pages/_app.js:142:1)
    at __webpack_require__ (/Users/***/frontend/web/.next/server/webpack-runtime.js:33:42) 
```

**Reply by osdiab** [reply]
i too experience this!

**Reply by arbejd** [reply]
Same, Identical error with d3-interpolate and d3-color.

**Reply by sokra** [reply]
I was not able to reproduce it with a fresh `create-next-app` created next.js app.

I tried

```
import * as ns from 'd3-scale';
console.log(ns);
```

in `pages/_app.js`. And the same with `d3` directly. Both work. With and without `esmExternals: true`.

Maybe try to upgrade some dependencies like `next` or `d3`. Otherwise, a repro would be great.

**Reply by Cyral** [reply]
@sokra Not sure if it matters but I import d3 like `import {scaleLinear} from 'd3-scale';` I will try to make a repro soon

**Reply by osdiab** [reply]
@sokra my issue is in using `react-simple-maps`, which imports `d3` itself, then fails. will see if i can make an example minimal repro.

**Reply by osdiab** [reply]
oh looks like i just needed up update all the `d3` dependencies, alls good now without any config!

**Comment by ThePaulMcBride** [comment]
There are some npm modules that only export using ems syntax, but are not marked as such. [AVA](https://github.com/avajs/ava) is one such example.

So far this feature works great, but it would be fantastic it be able to manually specify modules you know to need transpilled.

**Comment by rinarakaki** [comment]
Is it still far from here to solve #9607 and #23725?  It would help me use webpack loaders written in ESM in `next.config.js`.

**Reply by kimbaudi** [reply]
this is what I want as well. the ability to use ESM in `next.config.js` so my custom webpack loaders can use esm `import`/`export` instead of cjs `require`/`module.export`. Many npm packages we use are moving to pure ESM. I hope the next.js team can make this possible in the near future 🙏. Gatsby.js is looking into support ESM as well (see https://github.com/gatsbyjs/gatsby/discussions/31599#discussioncomment-1275280)

**Comment by lunelson** [comment]
This feature helps with a few use-cases, for example ESM modules used inside of page data-fetching methods; but we really still need 

1. the ability to mark modules for babel-transpiling as well, and 
2. to mark modules _explicitly as ESM_ if their `package.json` meta isn't correct

**Reply by kimbaudi** [reply]
> This feature helps with a few use-cases, for example ESM modules used inside of page data-fetching methods;

I am on next.js v11.1.0 and I am still able to use ESM modules inside of page data-fetching methods without the `experimental.esmExternals: true` feature.
For example, I can use [unified](https://www.npmjs.com/package/unified) ESM only package to parse markdown files in my `getStaticProps` method in a file under my `pages` directory.
So I'm not sure what this flag is supposed to do exactly.

**Comment by kimbaudi** [comment]
I'm using next.js v11.1.0 and the `experimental.esmExternals: true` flag does not do anything for me. Without this flag enabled, I am still able to use esm (`import`/`export`) in `pages` directory to prefetch data using `getStaticProps`. I thought I need to enable this flag to import esm only npm packages such as `unified`, `remark-parse`, etc. inside `pages` directory, but i can still import esm only npm packages without it. Also, I was hoping I can use esm (`import`/`export`) statement instead of cjs (`require` & `module.export`) in `next.config.js`, but this obviously won't work since `next.config.js` is cjs file. My feedback/request is for Next.js to be pure ESM package so that I can use esm (`import`/`export`) from `next.config.js`

**Comment by mattpodolak** [comment]
I enabled this experimental feature, but was unable to dynamically import ESM modules from my file system from the Next.js api, will this capability be added?

When importing this error was thrown: `SyntaxError: Unexpected token 'export'`

**Comment by imaksp** [comment]
If I use 'module' field instead of 'main' from package.json by extending webpack config I am getting following error with popperjs: 
```bash
node_modules/@popperjs/core/lib/modifiers/arrow.js:1
import getBasePlacement from "../utils/getBasePlacement.js";
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:988:16)
    at Module._compile (internal/modules/cjs/loader.js:1036:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12)
    at Module.require (internal/modules/cjs/loader.js:961:19)
    at require (internal/modules/cjs/helpers.js:92:18)
```

Also is there any way to force esm import for specific lib (without updating webpack config), if lib supports both?

**Comment by jimisaacs** [comment]
anyone else getting an error when importing `nanoid`
```
Error: Not supported

This error happened while generating the page. Any console logs will be displayed in the terminal window.
Call Stack
Object.nanoid
{redacted}/_app.js (2452:1)
__webpack_require__
{redacted}/.next/server/webpack-runtime.js (33:42)
eval
...{redacted}
```

**Comment by ctjlewis** [comment]
Some notes, having been exploring this recently:

## Do not add `"type": "module"` to your `package.json` (yet)

The `esmExternals` experimental flag only adds support for ESM *externals*, i.e., you can import ESM dependencies. You cannot bundle your Next project as an ES module yet.

The same reason that Next cannot load an ESM config (see #9607) is the same reason it cannot load your project if you set `"type": "module"`. If you mark your project as an ES module with `"type": "module"`, Next will break while trying to load `next.config.js` and compiled JS in `.next`—these are currently loaded via hard-coded `require(...)` statements, and trying to `require()` an ES module throws `ERR_REQUIRE_ESM`:

https://github.com/vercel/next.js/blob/16a737d9b55923c1ef7f89b99b3ef39e0d8e8488/packages/next/server/config.ts#L473

It will actually take upgrading Next core itself to ESM to add full ESM support (as widely requested, e.g. in #9607), which I will explain in detail in a separate issue. It is worth it to undertake that effort, but for now it will suffice to fix the resolution issues below IMO, and let Next back-transpile ESM to CJS for deploy.

## ESM re-export resolution

There seem to be some edge cases with export resolution that break Webpack, particularly when circularly re-importing something from an index.

**Work around this by importing members from the module they were originally exported from**, i.e. `import { something } from '../context/something'` instead of importing a re-export via `import { something } from '..'`, or add the following to your Next config:

```js
// next.config.js
{
  // ...
  webpack: (config) => {
    config.optimization.providedExports = true
    return config;
  }
}
```

See #29361.

**Comment by steve-taylor** [comment]
This setting works only in the browser. Is there a way to make it work server-side as well?

**Comment by giovannibenussi** [comment]
I'm having issues because directory imports. We have some old components that uses Apollo's `Query` and `Mutation` that [need to be imported like this](https://www.apollographql.com/docs/react/api/react/components/#mutation):

```
import { Query } from '@apollo/client/react/components'
```

I tried running `NODE_OPTIONS="--experimental-specifier-resolution=node" yarn dev`, but it doesn't work.

**Comment by eric-burel** [comment]
Hi guys, I am trying to build a package for Next.js, using TS and ESM. Dynamic import of client-only code doesn't seem to work.
```
Module not found: ESM packages (my-package-esbuild/client) need to be imported. Use 'import' to reference the package instead. https://nextjs.org/docs/messages/import-esm-externals
```

The package seems to be built correctly, but maybe this is due to TS interfering in the Next app, transforming import() into require()?

Repro: https://github.com/VulcanJS/npm-the-right-way, run `yarn run publish` in the Esbuild package, and install and run the demo next app, open the Esbuild demo

**Comment by fabb** [comment]
I have an issue with [`image-blob-reduce`](https://www.npmjs.com/package/image-blob-reduce). It works fine in dev mode (the ES Module is used), but in the prod build, the ES Module is **not** used and a runtime error occurs because of some weird way webpack converts the `require`s in the non-esm code.

See https://github.com/nodeca/pica/issues/228 for some more context.

Any idea what could cause webpack to not use the existing ES Module for the prod build?

**Reply by Naes0** [reply]
I'm having a similar issue, I've bundled up a package locally using rollup exporting both cjs and esm files. 

It worked fine locally (referencing the es module file) via npm link, but then when I deployed my package and installed it, it won't use the esm file. It keeps trying to use the cjs which is causing issues.

Did you have any updates?

**Reply by eric-burel** [reply]
Yes similar issue there https://github.com/vercel/next.js/issues/35112
On the contrary, there are issue when using link instead of the published package https://github.com/vercel/next.js/issues/35110 https://github.com/vercel/next.js/issues/34956

**Comment by chentsulin** [comment]
For now, `esmExternals` doesn't work well with `styled-components` due to https://github.com/styled-components/styled-components/issues/3601 .

**Comment by The-Code-Monkey** [comment]
```
Error [ERR_UNSUPPORTED_DIR_IMPORT]: Directory import '/Users/iv0706/code/TechStack/platforms/react-cms/node_modules/@aw-web-design/components/dist/esm/primatives' is not supported resolving ES modules imported from /Users/iv0706/code/TechStack/platforms/react-cms/node_modules/@aw-web-design/components/dist/esm/index.js
```

I seem to be getting a similar error as above running next 12

**Reply by karol-f** [reply]
Maybe use `babel-plugin-direct-import` (but if you use SWC and not Babel it's not a solution for you). 

Problematic package should export CJS version of package so node won't brek like this.

Solution for now - use https://github.com/martpie/next-transpile-modules for problematic package.

**Comment by imaksp** [comment]
is this still experimental, when this will be available outside experimental config ?

**Reply by karol-f** [reply]
> is this still experimental, when this will be available outside experimental config ?

It's default from Next.js 12

**Reply by imaksp** [reply]
I don't think so, it is still allowed without warning in experimental config, & value is still being used from this experimental config here:
https://github.com/vercel/next.js/blob/canary/packages/next/src/build/webpack-config.ts#L515

**Reply by karol-f** [reply]
Here is info about it https://nextjs.org/blog/next-12#es-modules-support-and-url-imports (with link to 11.1 announcement of esmExternals)
